category_theory/limits/preserves.lean

```diff
@@ -0,0 +1,180 @@
+-- Copyright (c) 2018 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison, Reid Barton
+
+-- Preservation and reflection of (co)limits.
+
+import category_theory.whiskering
+import category_theory.limits.limits
+
+open category_theory
+
+namespace category_theory.limits
+
+universes u₁ u₂ u₃ v
+
+variables {C : Type u₁} [𝒞 : category.{u₁ v} C]
+variables {D : Type u₂} [𝒟 : category.{u₂ v} D]
+include 𝒞 𝒟
+
+variables {J : Type v} [small_category J] {K : J ⥤ C}
+
+/- Note on "preservation of (co)limits"
+
+There are various distinct notions of "preserving limits". The one we
+aim to capture here is: A functor F : C → D "preserves limits" if it
+sends every limit cone in C to a limit cone in D. Informally, F
+preserves all the limits which exist in C.
+
+Note that:
+
+* Of course, we do not want to require F to *strictly* take chosen
+  limit cones of C to chosen limit cones of D. Indeed, the above
+  definition makes no reference to a choice of limit cones so it makes
+  sense without any conditions on C or D.
+
+* Some diagrams in C may have no limit. In this case, there is no
+  condition on the behavior of F on such diagrams. There are other
+  notions (such as "flat functor") which impose conditions also on
+  diagrams in C with no limits, but these are not considered here.
+
+In order to be able to express the property of preserving limits of a
+certain form, we say that a functor F preserves the limit of a
+diagram K if F sends every limit cone on K to a limit cone. This is
+vacuously satisfied when K does not admit a limit, which is consistent
+with the above definition of "preserves limits".
+
+-/
+
+class preserves_limit (K : J ⥤ C) (F : C ⥤ D) :=
+(preserves : Π {c : cone K}, is_limit c → is_limit (F.map_cone c))
+class preserves_colimit (K : J ⥤ C) (F : C ⥤ D) :=
+(preserves : Π {c : cocone K}, is_colimit c → is_colimit (F.map_cocone c))
```

Comment:
These should be marked `Prop`. More generally you should always write the target type of a definition (such as the ones below), for documentation purposes. There are possibly exceptions but they should be few.

Reply:
These are actually intentionally not `Prop`s, because `is_(co)limit` are not Props either. I guess we could write subsingleton instances for all these `preserves_*` classes...

Reply:
Okay. In that case it's even more important to give the type, so it doesn't look like an oversight.

Reply:
Okay, I've annotated these all with `Type X`, for an appropriate universe `X` (it's different for `preserves_limit` and for `preserves_limits`!).

Reply:
I wrote `subsingleton` instances for `preserves_*` (but not yet for `reflects_*`). They are a bit gross, however, so if someone wants to improve or drop them that's fine.

(In particular, `preserves_limits_of_shape_subsingleton` doesn't seem to actually be found by `apply_instance`, when I need it for `preserves_limits_subsingleton`.

Reply:
You just need to add `resetI` after introducing `𝒥` by `funext`, so that it is available to instance search.

It should also be possible to use `pi.subsingleton` for the ones which are Pi types, but the elaborator does not seem to want to make that easy.

Reply:
Thanks! I fixed those proofs, and added instances for `reflects_*`.